File: SEPADirectDebit/src/main/java/com/braintreepayments/api/SEPADirectDebitApi.java
package com.braintreepayments.api;

import androidx.annotation.VisibleForTesting;

import org.json.JSONException;
import org.json.JSONObject;

import javax.net.ssl.SSLException;
import javax.net.ssl.SSLSocketFactory;

class SEPADirectDebitApi {

    private final BraintreeClient braintreeClient;

    SEPADirectDebitApi(BraintreeClient braintreeClient) { this.braintreeClient = braintreeClient; }

    void createMandate(SEPADirectDebitRequest sepaDirectDebitRequest, String returnUrlScheme, final CreateMandateCallback callback) {
        try {
            JSONObject jsonObject = buildCreateMandateRequest(sepaDirectDebitRequest, returnUrlScheme);
            String url = "v1/sepa_debit";
            braintreeClient.sendPOST(url, jsonObject.toString(), new HttpResponseCallback() {

                @Override
                public void onResult(String responseBody, Exception httpError) {
                    if (responseBody != null) {
                        try {
                            CreateMandateResult result = parseCreateMandateResponse(responseBody);
                            callback.onResult(result, null);
                        } catch (JSONException e) {
                            callback.onResult(null, e);
                        }
                    } else if (httpError != null) {
                        callback.onResult(null, httpError);
                    }
                }
            });
        } catch (JSONException e) {
            callback.onResult(null, e);
        }
    }

    void tokenize(String ibanLastFour, String customerId, String bankReferenceToken, String mandateType, final SEPADirectDebitTokenizeCallback callback) {
        try {
            JSONObject jsonObject = buildTokenizeRequest(ibanLastFour, customerId, bankReferenceToken, mandateType);
            String url = "v1/payment_methods/sepa_debit_accounts";
            braintreeClient.sendPOST(url, jsonObject.toString(), new HttpResponseCallback() {

                @Override
                public void onResult(String responseBody, Exception httpError) {
                   if (responseBody != null) {
                       try {
                           SEPADirectDebitNonce nonce = parseTokenizeResponse(responseBody);
                           callback.onResult(nonce, null);
                       } catch (JSONException jsonException) {
                           callback.onResult(null, jsonException);
                       }
                   } else if (httpError != null) {
                       callback.onResult(null, httpError);
                   }
                }
            });
        } catch (JSONException e) {
            callback.onResult(null, e);
        }
    }

    private SEPADirectDebitNonce parseTokenizeResponse(String responseBody) throws JSONException {
        JSONObject jsonResponse = new JSONObject(responseBody);
        return SEPADirectDebitNonce.fromJSON(jsonResponse);
    }

    private JSONObject buildTokenizeRequest(String ibanLastFour, String customerId, String bankReferenceToken, String mandateType) throws JSONException {
        JSONObject accountData = new JSONObject()
                .put("last_4", ibanLastFour)
                .put("merchant_or_partner_customer_id", customerId)
                .put("bank_reference_token", bankReferenceToken)
                .put("mandate_type", mandateType);
        JSONObject requestData = new JSONObject()
                .put("sepa_debit_account", accountData);

        return requestData;
    }

    private CreateMandateResult parseCreateMandateResponse(String responseBody) throws JSONException {
        JSONObject json = new JSONObject(responseBody);
        JSONObject sepaDebitAccount = json.getJSONObject("message").getJSONObject("body").getJSONObject("sepaDebitAccount");
        String approvalUrl = sepaDebitAccount.getString("approvalUrl");
        String ibanLastFour = sepaDebitAccount.getString("last4");
        String customerId = sepaDebitAccount.getString("merchantOrPartnerCustomerId");
        String bankReferenceToken = sepaDebitAccount.getString("bankReferenceToken");
        String mandateType = sepaDebitAccount.getString("mandateType");

        return new CreateMandateResult(approvalUrl, ibanLastFour, customerId, bankReferenceToken, mandateType);
    }

    private JSONObject buildCreateMandateRequest(SEPADirectDebitRequest sepaDirectDebitRequest, String returnUrlScheme) throws JSONException {
        JSONObject sepaDebitData = new JSONObject()
            .putOpt("account_holder_name", sepaDirectDebitRequest.getAccountHolderName())
            .putOpt("merchant_or_partner_customer_id", sepaDirectDebitRequest.getCustomerId())
            .putOpt("iban", sepaDirectDebitRequest.getIban())
            .putOpt("mandate_type", sepaDirectDebitRequest.getMandateType().toString());

        if (sepaDirectDebitRequest.getBillingAddress() != null) {
            JSONObject billingAddress = new JSONObject()
                    .putOpt("address_line_1", sepaDirectDebitRequest.getBillingAddress().getStreetAddress())
                    .putOpt("address_line_2", sepaDirectDebitRequest.getBillingAddress().getExtendedAddress())
                    .putOpt("admin_area_1", sepaDirectDebitRequest.getBillingAddress().getLocality())
                    .putOpt("admin_area_2", sepaDirectDebitRequest.getBillingAddress().getRegion())
                    .putOpt("postal_code", sepaDirectDebitRequest.getBillingAddress().getPostalCode())
                    .putOpt("country_code", sepaDirectDebitRequest.getBillingAddress().getCountryCodeAlpha2());

            sepaDebitData.put("billing_address", billingAddress);
        }

        String cancelUrl = String.format("%s://sepa/cancel", returnUrlScheme);
        String successUrl = String.format("%s://sepa/success", returnUrlScheme);

        JSONObject requestData = new JSONObject()
                .put("sepa_debit", sepaDebitData)
                .put("cancel_url", cancelUrl)
                .put("return_url", successUrl);

        if (sepaDirectDebitRequest.getMerchantAccountId() != null) {
            requestData.putOpt("merchant_account_id", sepaDirectDebitRequest.getMerchantAccountId());
        }

        return requestData;
    }
}

File: SEPADirectDebit/src/main/java/com/braintreepayments/api/sepadirectdebit/SEPADirectDebitApi.kt
package com.braintreepayments.api.sepadirectdebit

import com.braintreepayments.api.core.BraintreeClient
import com.braintreepayments.api.sepadirectdebit.SEPADirectDebitNonce.Companion.fromJSON
import org.json.JSONException
import org.json.JSONObject

internal class SEPADirectDebitApi(
    private val braintreeClient: BraintreeClient
) {

    suspend fun createMandate(
        sepaDirectDebitRequest: SEPADirectDebitRequest,
        returnUrlScheme: String
    ): CreateMandateResult {
        val jsonObject = buildCreateMandateRequest(sepaDirectDebitRequest, returnUrlScheme)
        val url = "/v1/sepa_debit"
        val responseBody = braintreeClient.sendPOST(url, jsonObject.toString())
        return parseCreateMandateResponse(responseBody)
    }

    suspend fun tokenize(
        ibanLastFour: String,
        customerId: String,
        bankReferenceToken: String,
        mandateType: String
    ): SEPADirectDebitNonce {
        val jsonObject = buildTokenizeRequest(ibanLastFour, customerId, bankReferenceToken, mandateType)
        val url = "/v1/payment_methods/sepa_debit_accounts"
        val responseBody = braintreeClient.sendPOST(url, jsonObject.toString())
        return parseTokenizeResponse(responseBody)
    }

    @Throws(JSONException::class)
    private fun parseTokenizeResponse(responseBody: String): SEPADirectDebitNonce {
        val jsonResponse = JSONObject(responseBody)
        return fromJSON(jsonResponse)
    }

    @Throws(JSONException::class)
    private fun buildTokenizeRequest(
        ibanLastFour: String, customerId: String,
        bankReferenceToken: String, mandateType: String
    ): JSONObject {
        val accountData = JSONObject()
            .put("last_4", ibanLastFour)
            .put("merchant_or_partner_customer_id", customerId)
            .put("bank_reference_token", bankReferenceToken)
            .put("mandate_type", mandateType)
        val requestData = JSONObject()
            .put("sepa_debit_account", accountData)

        return requestData
    }

    @Throws(JSONException::class)
    private fun parseCreateMandateResponse(responseBody: String): CreateMandateResult {
        val json = JSONObject(responseBody)
        val sepaDebitAccount = json.getJSONObject("message").getJSONObject("body")
            .getJSONObject("sepaDebitAccount")
        val approvalUrl = sepaDebitAccount.getString("approvalUrl")
        val ibanLastFour = sepaDebitAccount.getString("last4")
        val customerId = sepaDebitAccount.getString("merchantOrPartnerCustomerId")
        val bankReferenceToken = sepaDebitAccount.getString("bankReferenceToken")
        val mandateType = sepaDebitAccount.getString("mandateType")

        return CreateMandateResult(
            approvalUrl,
            ibanLastFour,
            customerId,
            bankReferenceToken,
            SEPADirectDebitMandateType.valueOf(mandateType)
        )
    }

    @Throws(JSONException::class)
    private fun buildCreateMandateRequest(
        sepaDirectDebitRequest: SEPADirectDebitRequest,
        returnUrlScheme: String
    ): JSONObject {
        val sepaDebitData = JSONObject()
            .putOpt("account_holder_name", sepaDirectDebitRequest.accountHolderName)
            .putOpt("merchant_or_partner_customer_id", sepaDirectDebitRequest.customerId)
            .putOpt("iban", sepaDirectDebitRequest.iban)
            .putOpt("mandate_type", sepaDirectDebitRequest.mandateType.toString())

        val requestBillingAddress = sepaDirectDebitRequest.billingAddress
        if (requestBillingAddress != null) {
            val billingAddress = JSONObject()
                .putOpt(
                    "address_line_1",
                    requestBillingAddress.streetAddress
                )
                .putOpt(
                    "address_line_2",
                    requestBillingAddress.extendedAddress
                )
                .putOpt(
                    "admin_area_1",
                    requestBillingAddress.locality
                )
                .putOpt("admin_area_2",
                    requestBillingAddress.region
                )
                .putOpt(
                    "postal_code",
                    requestBillingAddress.postalCode
                )
                .putOpt(
                    "country_code",
                    requestBillingAddress.countryCodeAlpha2
                )

            sepaDebitData.put("billing_address", billingAddress)
        }

        val cancelUrl = "$returnUrlScheme://sepa/cancel"
        val successUrl = "$returnUrlScheme://sepa/success"

        val requestData = JSONObject()
            .put("sepa_debit", sepaDebitData)
            .put("cancel_url", cancelUrl)
            .put("return_url", successUrl)

        if (sepaDirectDebitRequest.merchantAccountId != null) {
            requestData.putOpt(
                "merchant_account_id",
                sepaDirectDebitRequest.merchantAccountId
            )
        }

        if (sepaDirectDebitRequest.locale != null) {
            requestData.putOpt("locale", sepaDirectDebitRequest.locale)
        }

        return requestData
    }
}
